test/optimization/test_quadratic_program.py

```diff
@@ -66,6 +66,144 @@ def test_name_setter(self):
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p = QuadraticProgram()
+            q_p.binary_var_dict(key_format='{}{}')
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p = QuadraticProgram()
+            q_p.binary_var_dict(keys=0)
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p = QuadraticProgram()
+            q_p.binary_var_dict(name='a')
+            q_p.binary_var_dict(name='a')
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p = QuadraticProgram()
+            q_p.binary_var_dict(key_format='_{{}}')
+
+    def test_var_list(self):
+        """test {binary,integer,continuous}_var_list"""
+        q_p = QuadraticProgram()
+
+        c_count = 0
+        b_count = 0
+        i_count = 0
+
+        def verify_counts():
+            self.assertEqual(q_p.get_num_vars(), c_count + b_count + i_count)
+            self.assertEqual(q_p.get_num_continuous_vars(), c_count)
+            self.assertEqual(q_p.get_num_binary_vars(), b_count)
+            self.assertEqual(q_p.get_num_integer_vars(), i_count)
+
+        def check_list(var_list, offset):
+            verify_counts()
+            for i, x in enumerate(var_list):
+                y = q_p.get_variable(i + offset)
+                z = q_p.get_variable(x.name)
+                self.assert_equal(x, y)
+                self.assert_equal(x, z)
+
+        d_0 = q_p.continuous_var_list(name='a', key_format='_{}', keys=3)
+        c_count += 3
+        check_list(d_0, 0)
+
+        d_1 = q_p.binary_var_list(name='b', keys=5)
+        b_count += 5
+        check_list(d_1, len(d_0))
+
+        d_2 = q_p.integer_var_list(key_format='_{}', keys=7, lowerbound=-4, upperbound=10)
+        i_count += 7
+        check_list(d_2, len(d_0) + len(d_1))
+
+        d_3 = q_p.continuous_var_list(name='a', key_format='_{}', keys=3)
+        c_count += 3
+        check_list(d_3, len(d_0) + len(d_1) + len(d_2))
+
+        d_4 = q_p.continuous_var_list(name='c', keys=range(3))
+        c_count += 3
+        check_list(d_4, len(d_0) + len(d_1) + len(d_2) + len(d_3))
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p.binary_var_list(name='c0')
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p.binary_var_list(name='c', keys=range(3))
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p = QuadraticProgram()
+            q_p.binary_var_list(key_format='{}{}')
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p = QuadraticProgram()
+            q_p.binary_var_list(keys=0)
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p = QuadraticProgram()
+            q_p.binary_var_list(name='a')
+            q_p.binary_var_list(name='a')
+
+        with self.assertRaises(QiskitOptimizationError):
+            q_p = QuadraticProgram()
+            q_p.binary_var_list(key_format='_{{}}')
+
```

Comment:
I'd add a test where all methods of adding variables (list, dict, one by one) are called randomly.